### PR TITLE
fix: [Search v1] Search page scroll is not smooth

### PR DESCRIPTION
**Pull request by charles-liang**
@eh2077 

### Details

[#41581 [$250] [Search v1] Search page scroll is not smooth](https://github.com/Expensify/App/issues/41581#top)

### Fixed Issues

$ https://github.com/Expensify/App/issues/41581#top
PROPOSAL: https://github.com/Expensify/App/issues/41581#issuecomment-2097784758

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open the search page, Profile > Troubleshoot > New Search Page
2. Scroll through the list of results
3. Notice that the items flicker and on mobile the list jumps around

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[untitled.webm](https://github.com/Expensify/App/assets/13301734/7029c003-8e1f-4396-bff2-640b355a200d)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/13301734/ef1d1c6a-472d-4818-9e96-015c067ca775

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/13301734/62ee86de-3b83-4840-b134-e681e9b0d346

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/13301734/c893a00a-6f82-4184-b9bf-e2ff4d9f825b

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/13301734/5eef35ff-43dc-436d-aef5-5a0d535a9e40

https://github.com/Expensify/App/assets/13301734/b08b629c-9388-4b7e-985a-349e6b27d1d8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/13301734/d3deef0d-9ab8-44cf-a6ca-ad8ab7a69075

https://github.com/Expensify/App/assets/13301734/9f473f93-1f8c-4d6d-b8fb-59640ef4d9bf

</details>

### CODE REVIEW

**Comment by eh2077** [comment]
> @eh2077 I compiled to Android native and iOS native, but I don't know how to switch the .env file. I tried hardcoding isDeveloping to open the test menu, entered the Search Page, but couldn't load the data. Can you tell me how I should test on these two platforms, or if it's not necessary to test right now?

@charles-liang Do you mean trying to open the search page on native Apps? If so, I think you can just follow the first reproduction step from the original issue
> Open the search page, Profile > Troubleshoot > New Search Page

**Comment by eh2077** [comment]
@charles-liang Can you handle the conflicts?

**Comment by eh2077** [comment]
@charles-liang And don’t worry if CI workflows are not running automatically. It needs help from @luacmartins to trigger it because it's your first PR and it's a little bit special.

You can run `lint`, `prettier` and `typecheck` locally before pushing.

**Comment by eh2077** [comment]
I don't think it's necessary to edit `.env` file. Can you follow 
- https://github.com/expensify/app?tab=readme-ov-file#running-the-android-app-
- https://github.com/expensify/app?tab=readme-ov-file#running-the-ios-app-

to run native apps?

**Comment by eh2077** [comment]
@charles-liang When opening a PR, please merge your branch to the `main` branch` as stated [here](https://github.com/expensify/app?tab=readme-ov-file#predeploy)

**Comment by eh2077** [comment]
> I mean i run it in iOS And android native will use production env and cannot enter in search page because it only open in dev or staging I don't know how to change it in the script.

I think the issue you encountered here is because you're not working on the `main` branch`  which contains default settings of `dev` env.

File: src/components/SelectionList/BaseSelectionList.tsx
Comment (eh2077) [inline src/components/SelectionList/BaseSelectionList.tsx:366]: Why do we need this change?

File: src/components/Search.tsx
Comment (eh2077) [reply src/components/Search.tsx:75]: Should we use different values for desktop and mobile devices? As their screens size are different

**Comment by eh2077** [comment]
I'll continue testing when the site is full recovered.

**Comment by eh2077** [comment]
@charles-liang I'm checkout this branch locally and still can't open search page.

image

**Comment by eh2077** [comment]
I also tested on the main branch and it was same - loading. But the staging env is working. So, I'm going to revisit this after the merging freeze.

**Comment by trjExpensify** [comment]
@eh2077 can you open the console and head to the `Network` tab before you go to search page. Then look for the `Search` command and check if you're getting errors. I wonder if you're running into something I did recently with a 501 for an invalid tax percentage.

**Comment by eh2077** [comment]
@trjExpensify Oh, sorry I forgot to attach the screenshot of request. Yeah, the back return invalid data though the HTTP code is 200.

image

cc @charles-liang

**Comment by eh2077** [comment]
> So？ what do I need to do? I feel like it's not an issue with my code.

@charles-liang In this case, I think you can just wait until I'm able to re-test it and then provide feedback. Thanks.

@trjExpensify Should we pull in another internal engineer (since @luacmartins is OOO) to take a look at the API issue, or should we just wait a little while?

**Comment by luacmartins** [comment]
@eh2077 @charles-liang the pagination API is only in staging and we have a merge freeze in place. We'll need to wait a bit longer to get that deployed to production. Meanwhile, you should be able to test this using the staging API. Let me know if you still have issues with that.

File: src/components/Search.tsx
Comment (luacmartins) [reply src/components/Search.tsx:85]: can we leave a comment explaining that please?

**Review by luacmartins — COMMENTED** [review]
LGTM, left a small comment

**Comment by luacmartins** [comment]
@charles-liang lint is failing too

File: src/components/SelectionList/TransactionListItem.tsx
Comment (eh2077) [inline src/components/SelectionList/TransactionListItem.tsx:185]: Can you combine `PureComponent` and `memo` like
```js
const Greeting = memo(function Greeting({ name }) {
  console.log("Greeting was rendered at", new Date().toLocaleTimeString());
  return <h3>Hello{name && ', '}{name}!</h3>;
}, arePropsEqual);
```

see https://react.dev/reference/react/PureComponent

File: src/components/SelectionList/types.ts
Comment (eh2077) [inline src/components/SelectionList/types.ts:417]: Can we remove this?

**Comment by eh2077** [comment]
@charles-liang It's still not smooth to scroll on mobile Chrome. 


https://github.com/Expensify/App/assets/117511920/5bce7b6f-9a39-4c29-bece-75702f2ddeac

**Comment by eh2077** [comment]
> > @charles-liang It's still not smooth to scroll on mobile Chrome.
> > 
> >   
> >     
> >     
> > 
> >     Screen.Recording.2024-05-21.at.11.31.52.PM.mov
> 
> may be we should decrease the update frequency for android mweb( increasing batch period from 200ms to 400ms or more) or decrease the effect to decrease the render cost for android mweb

It looks better after restarting my emulator. 

https://github.com/Expensify/App/assets/117511920/a5f01fb6-b81f-4737-a72b-ede15d524bb0

File: src/components/SelectionList/types.ts
Comment (eh2077) [reply src/components/SelectionList/types.ts:417]: I mean removing the unnecessary eslint-disable comments

**Review by eh2077 — CHANGES_REQUESTED** [review]
@charles-liang Please help to address comments and run 
- `npm run prettier`
- `npm run lint`
- `npm run typecheck`
- `npm run test`

before pushing.

**Comment by luacmartins** [comment]
@charles-liang please avoid force-pushing changes to the branch as that makes it a bit harder to review the code

**Comment by eh2077** [comment]
> Should I rebase my branch onto the main branch and then make changes? Or should I merge the latest code from the main branch into my branch?

@charles-liang You can follow this [SO](https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks/38955036#38955036). 

Retaining all commits is more friendly for collaboration - making it easier to review the code as @luacmartins mentioned.

File: src/components/SelectionList/TransactionListItem.tsx
Comment (eh2077) [inline src/components/SelectionList/TransactionListItem.tsx:83]:
```suggestion
const MerchantCell = memo(({showTooltip, transactionItem, merchant, description}: MerchantCellProps) => {
    const styles = useThemeStyles();
    return (
        <TextWithTooltip
            shouldShowTooltip={!!showTooltip}
            text={transactionItem.shouldShowMerchant ? merchant : description}
            style={[styles.optionDisplayName, styles.label, styles.pre, styles.justifyContentCenter]}
        />
    );
}, arePropsEqual);
```

**Review by eh2077 — CHANGES_REQUESTED** [review]
Please use arrow callback instead of turning off eslint.

File: src/components/SelectionList/TransactionListItem.tsx
Comment (eh2077) [inline src/components/SelectionList/TransactionListItem.tsx:54]: NAB: Though, personally, I don't prefer adding `Memo` prefix to the name. I think it's fine to just have `DateCell`.

File: src/components/SelectionList/BaseSelectionList.tsx
Comment (eh2077) [inline src/components/SelectionList/BaseSelectionList.tsx:364]: I think this type conflict is caused by this PR. Can you explain why?

File: src/components/Search.tsx
Comment (eh2077) [inline src/components/Search.tsx:85]: I think these comments can be better wording. You can try to improve it or I'll revisit and suggest changes.

File: src/components/SelectionList/types.ts
Comment (eh2077) [inline src/components/SelectionList/types.ts:385]: The default value is 5 right?

File: src/components/SelectionList/types.ts
Comment (eh2077) [inline src/components/SelectionList/types.ts:443]: Please remove this comment.

File: src/components/Search.tsx
Comment (eh2077) [inline src/components/Search.tsx:85]:
```suggestion
            // To enhance the smoothness of scrolling and minimize the risk of encountering blank spaces during scrolling,
            // we have configured a larger windowSize and a longer delay between batch renders.
            // The windowSize determines the number of items rendered before and after the currently visible items.
            // A larger windowSize helps pre-render more items, reducing the likelihood of blank spaces appearing.
            // The updateCellsBatchingPeriod sets the delay (in milliseconds) between rendering batches of cells.
            // A longer delay allows the UI to handle rendering in smaller increments, which can improve performance and smoothness.
            // For more information, refer to the React Native documentation:
            // https://reactnative.dev/docs/0.73/optimizing-flatlist-configuration#windowsize
            // https://reactnative.dev/docs/0.73/optimizing-flatlist-configuration#updatecellsbatchingperiod
            windowSize={111}
            updateCellsBatchingPeriod={200}
```

File: src/components/SelectionList/TransactionListItem.tsx
Comment (eh2077) [inline src/components/SelectionList/TransactionListItem.tsx:55]: We can omit the return type
```suggestion
    ({ showTooltip, date, isLargeScreenWidth }: DateCellProps) => {
```

**Review by eh2077 — CHANGES_REQUESTED** [review]

File: src/components/SelectionList/BaseSelectionList.tsx
Comment (eh2077) [reply src/components/SelectionList/BaseSelectionList.tsx:409]: OK. As the fix doesn’t look straightforward, I'm not sure if we should fix it here or just let it as it is (Maybe there's an ongoing fix somewhere). I'd like to defer this to @luacmartins

**Comment by eh2077** [comment]
@luacmartins This PR overall looks good to me and tested well. Just defer one [comment](https://github.com/Expensify/App/pull/42170#discussion_r1616236667) (about fixing type error from main) to you. Kindly help to trigger the workflow to see if CI is green, thanks!

**Comment by luacmartins** [comment]
Running workflow now.

**Comment by luacmartins** [comment]
@eh2077 seems like typecheck is passing.

File: src/components/SelectionList/TransactionListItem.tsx
Comment (luacmartins) [inline src/components/SelectionList/TransactionListItem.tsx:262]: We need the `styles.flex1` here otherwise the text overflows. I won't block on this though, since it's a simple change we can do in a follow up and it's more important to get this optimization out and resolve conflicts on ongoing PRs.

Screenshot 2024-05-27 at 12 25 26 PM

**Review by luacmartins — APPROVED** [review]

**Review by eh2077 — APPROVED** [review]

**Review by luacmartins — APPROVED** [review]

**Pull request merged**
